File: services/api/src/test/java/com/voltpilot/api/uems/BezugsPeriodeTest.java
package com.voltpilot.api.uems;

import static org.assertj.core.api.Assertions.assertThat;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.voltpilot.api.uems.BezugsPeriode.Periodendeutung;
import com.voltpilot.api.uems.BezugsPeriode.Zeitdeutung;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * Das PERIODEN-Modul (UEMS AP-09 IP-3) gegen die geteilte Vektor-Datei
 * {@code docs/contracts/v2/bezugsdaten-vectors.json} — Familien {@code periode}, {@code zeit},
 * {@code stunden} — und gegen die drei Fallen, an denen es richtig oder falsch wird: der Tag mit
 * 23 oder 25 Stunden, der mehrdeutige bzw. nicht existierende Zeitpunkt und der Unterschied
 * zwischen „passt nicht“ und „noch nicht zu Ende“.
 *
 * <p>Der Aufruf geht hier DIREKT an {@link BezugsPeriode}; {@code BezugsdatenVectorsTest} fährt
 * dieselben Fälle über den Anruf in {@link BezugsdatenRegeln}.
 */
class BezugsPeriodeTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static final Path VECTORS =
            Path.of("..", "..", "docs", "contracts", "v2", "bezugsdaten-vectors.json");

    private static final Path REGELN =
            Path.of("src", "main", "java", "com", "voltpilot", "api", "uems", "BezugsdatenRegeln.java");

    /** Die Zeitzone des Standorts der Referenzfälle (Ahrenberg). */
    private static final ZoneId AHRENBERG = ZoneId.of("Europe/Berlin");

    private static JsonNode vektoren() throws Exception {
        return MAPPER.readTree(Files.readString(VECTORS));
    }

    private static String text(JsonNode n) {
        return n.isMissingNode() || n.isNull() ? null : n.asText();
    }

    /** Jede Prüfung der Familien {@code periode}, {@code zeit} und {@code stunden} — am Modul. */
    @TestFactory
    List<DynamicTest> dieFamilienPeriodeZeitUndStunden() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode fall : vektoren().path("cases")) {
            for (JsonNode p : fall.path("pruefungen")) {
                String regel = p.path("regel").asText();
                if (!List.of("periode", "zeit", "stunden").contains(regel)) {
                    continue;
                }
                String name = fall.path("id").asText() + " · " + regel + " :: " + p.path("name").asText();
                tests.add(DynamicTest.dynamicTest(name, () -> pruefe(regel, name, p)));
            }
        }
        assertThat(tests).as("Prüfungen der drei Familien").hasSizeGreaterThanOrEqualTo(15);
        return tests;
    }

    private void pruefe(String regel, String name, JsonNode p) {
        JsonNode ein = p.path("eingang");
        JsonNode soll = p.path("ergebnis");
        switch (regel) {
            case "periode" -> {
                Periodendeutung ist = BezugsPeriode.periode(
                        text(ein.path("text")),
                        text(ein.path("von_text")),
                        text(ein.path("bis_text")),
                        ein.path("deutung").asText(),
                        ein.path("periode_art").asText(),
                        AHRENBERG,
                        ein.has("jetzt") ? BezugsPeriode.zeit(ein.path("jetzt").asText()) : null);
                assertThat(ist.schluessel()).as(name + " · schluessel").isEqualTo(text(soll.path("schluessel")));
                assertThat(ist.von() == null ? null : BezugsPeriode.iso(ist.von(), AHRENBERG))
                        .as(name + " · von")
                        .isEqualTo(text(soll.path("von")));
                assertThat(ist.bis() == null ? null : BezugsPeriode.iso(ist.bis(), AHRENBERG))
                        .as(name + " · bis")
                        .isEqualTo(text(soll.path("bis")));
                assertThat(ist.stunden())
                        .as(name + " · stunden")
                        .isEqualTo(soll.path("stunden").isNull() ? null : soll.path("stunden").asLong());
                assertThat(ist.befund()).as(name + " · befund").isEqualTo(text(soll.path("befund")));
            }
            case "zeit" -> {
                Zeitdeutung ist = BezugsPeriode.zeitpunkt(
                        ein.path("text").asText(),
                        ZoneId.of(ein.path("zeitzone").asText()),
                        text(ein.path("offset_in_datei")));
                assertThat(ist.zeitpunkt() == null ? null : BezugsPeriode.iso(ist.zeitpunkt(), AHRENBERG))
                        .as(name + " · zeitpunkt")
                        .isEqualTo(text(soll.path("zeitpunkt")));
                assertThat(ist.befund()).as(name + " · befund").isEqualTo(text(soll.path("befund")));
                List<String> varianten = new ArrayList<>();
                soll.path("varianten").forEach(x -> varianten.add(x.asText()));
                assertThat(ist.varianten()).as(name + " · varianten").isEqualTo(varianten);
            }
            default -> assertThat(BezugsPeriode.stundenDesTages(
                            LocalDate.parse(ein.path("tag").asText()), AHRENBERG))
                    .as(name)
                    .isEqualTo(soll.path("stunden").asLong());
        }
    }

    /** Die Kundensätze wohnen im Modul — und sind Wort für Wort die des Vertrags. */
    @Test
    void dieKundensaetzeSindDieDesVertrags() throws Exception {
        JsonNode saetze = vektoren().path("befund_saetze");
        BezugsPeriode.SAETZE.forEach((befund, satz) -> assertThat(satz)
                .as("Kundensatz " + befund)
                .isEqualTo(saetze.path(befund).asText()));
        assertThat(BezugsPeriode.SAETZE.keySet())
                .containsExactlyInAnyOrder(
                        BezugsPeriode.PERIODE_PASST_NICHT,
                        BezugsPeriode.PERIODE_NICHT_ZU_ENDE,
                        BezugsPeriode.ZEIT_MEHRDEUTIG,
                        BezugsPeriode.ZEIT_NICHT_VORHANDEN,
                        BezugsPeriode.DATUM_UNLESBAR);
    }

    /**
     * FALLE 1 — ein Tag hat nicht 24 Stunden, und die Stundenzahl wird bei der Verbrauchsregel
     * AP-08 BESTELLT statt hier ein zweites Mal gezählt.
     */
    @Test
    void derUmstellungstagHat25BeziehungsweiseTagHat23Stunden() {
        assertThat(BezugsPeriode.stundenDesTages(LocalDate.parse("2026-10-25"), AHRENBERG))
                .as("Rückstellung: der Tag ist eine Stunde LÄNGER")
                .isEqualTo(25);
        assertThat(BezugsPeriode.stundenDesTages(LocalDate.parse("2027-03-28"), AHRENBERG))
                .as("Vorstellung: der Tag ist eine Stunde KÜRZER")
                .isEqualTo(23);
        assertThat(BezugsPeriode.stundenDesTages(LocalDate.parse("2026-12-02"), AHRENBERG))
                .isEqualTo(24);

        // Nachgerechnet, nicht geglaubt: es ist dieselbe Zahl, die AP-08 nennt.
        for (String tag : List.of("2026-10-25", "2027-03-28", "2026-12-02")) {
            LocalDate d = LocalDate.parse(tag);
            assertThat(BezugsPeriode.stundenDesTages(d, AHRENBERG))
                    .as("dieselbe Zählung wie VerbrauchRegeln.stunden am " + tag)
                    .isEqualTo(VerbrauchRegeln.stunden(
                            d.atStartOfDay(AHRENBERG).toInstant(),
                            d.plusDays(1).atStartOfDay(AHRENBERG).toInstant()));
        }

        // Und die Periode erbt das: der Oktober 2026 hat 745 Stunden, nicht 744.
        Periodendeutung oktober = BezugsPeriode.periode(
                "2026-10", null, null, "periode", "monat", AHRENBERG, BezugsPeriode.zeit("2026-11-03T09:12:00+01:00"));
        assertThat(oktober.stunden()).isEqualTo(745);
        Periodendeutung maerz = BezugsPeriode.periode(
                "2027-03", null, null, "periode", "monat", AHRENBERG, BezugsPeriode.zeit("2027-04-02T09:00:00+02:00"));
        assertThat(maerz.stunden()).as("der März 2027 hat 743 Stunden").isEqualTo(743);
    }

    /**
     * FALLE 2 — der mehrdeutige und der nicht existierende Zeitpunkt sind BEFUNDE mit Kundensatz,
     * nie eine stille Wahl.
     */
    @Test
    void mehrdeutigUndNichtVorhandenSindBefundeMitKundensatz() {
        Zeitdeutung doppelt = BezugsPeriode.zeitpunkt("25.10.2026 02:30", AHRENBERG, null);
        assertThat(doppelt.zeitpunkt()).as("keine der beiden Möglichkeiten wird gewählt").isNull();
        assertThat(doppelt.befund()).isEqualTo(BezugsPeriode.ZEIT_MEHRDEUTIG);
        assertThat(doppelt.varianten())
                .as("BEIDE Möglichkeiten reisen mit, damit der Kunde wählen kann")
                .containsExactly("2026-10-25T02:30:00+02:00", "2026-10-25T02:30:00+01:00");
        assertThat(BezugsPeriode.satz(doppelt.befund()))
                .isEqualTo("Diesen Zeitpunkt gibt es an diesem Tag zweimal (Zeitumstellung). Geben Sie die Zone an.");

        Zeitdeutung fehlt = BezugsPeriode.zeitpunkt("28.03.2027 02:30", AHRENBERG, null);
        assertThat(fehlt.zeitpunkt()).as("nie auf 03:30 verschoben").isNull();
        assertThat(fehlt.befund()).isEqualTo(BezugsPeriode.ZEIT_NICHT_VORHANDEN);
        assertThat(fehlt.varianten()).isEmpty();
        assertThat(BezugsPeriode.satz(fehlt.befund()))
                .isEqualTo("Diesen Zeitpunkt gibt es an diesem Tag nicht (Zeitumstellung).");

        // Ein Offset in der Datei gewinnt immer und macht dieselbe Zeile eindeutig (Z5).
        Zeitdeutung mitOffset = BezugsPeriode.zeitpunkt("25.10.2026 02:30", AHRENBERG, "+02:00");
        assertThat(mitOffset.befund()).isNull();
        assertThat(BezugsPeriode.iso(mitOffset.zeitpunkt(), AHRENBERG)).isEqualTo("2026-10-25T02:30:00+02:00");
    }

    /**
     * FALLE 3 — „passt nicht“ (wird NIE geteilt) und „noch nicht zu Ende“ (kommt später wieder)
     * sind zwei Befunde mit zwei Sätzen, nicht einer.
     */
    @Test
    void passtNichtIstNichtDasselbeWieNochNichtZuEnde() {
        Instant jetzt = BezugsPeriode.zeit("2026-11-03T09:12:00+01:00");

        Periodendeutung woche =
                BezugsPeriode.periode(null, "28.09.2026", "04.10.2026", "von_bis", "monat", AHRENBERG, jetzt);
        assertThat(woche.befund()).isEqualTo(BezugsPeriode.PERIODE_PASST_NICHT);
        assertThat(woche.schluessel()).as("nichts wird geteilt, nichts nach Mehrheit zugeordnet").isNull();
        assertThat(woche.von()).isNull();
        assertThat(woche.bis()).isNull();
        assertThat(BezugsPeriode.satz(woche.befund()))
                .isEqualTo("Der gelieferte Zeitraum ist keine Periode dieser Bezugsgröße.");

        // Dieselbe Woche in einer WOCHEN-Bezugsgröße passt — die Reihe entscheidet, nicht der Text.
        Periodendeutung inWochenreihe =
                BezugsPeriode.periode(null, "28.09.2026", "04.10.2026", "von_bis", "woche", AHRENBERG, jetzt);
        assertThat(inWochenreihe.schluessel()).isEqualTo("2026-W40");
        assertThat(inWochenreihe.befund()).isNull();

        Periodendeutung laufend =
                BezugsPeriode.periode("2026-11", null, null, "periode", "monat", AHRENBERG, jetzt);
        assertThat(laufend.befund()).isEqualTo(BezugsPeriode.PERIODE_NICHT_ZU_ENDE);
        assertThat(BezugsPeriode.satz(laufend.befund())).isEqualTo("Diese Periode ist noch nicht zu Ende.");

        assertThat(BezugsPeriode.satz(BezugsPeriode.PERIODE_PASST_NICHT))
                .as("zwei Befunde, zwei Sätze — der Kunde tut zwei verschiedene Dinge")
                .isNotEqualTo(BezugsPeriode.satz(BezugsPeriode.PERIODE_NICHT_ZU_ENDE));

        // Und der laufende Monat wird nicht abgelehnt, weil er falsch wäre: einen Tag nach seinem
        // Ende nimmt dieselbe Reihe denselben Text an.
        Periodendeutung spaeter = BezugsPeriode.periode(
                "2026-11", null, null, "periode", "monat", AHRENBERG, BezugsPeriode.zeit("2026-12-01T00:00:00+01:00"));
        assertThat(spaeter.befund()).isNull();
        assertThat(spaeter.schluessel()).isEqualTo("2026-11");
    }

    /** E7/Z5 — die Zone kommt vom STANDORT; eine Vorlage darf davon abweichen. */
    @Test
    void dieZonenloseOrtszeitLiestDieZonedesStandorts() {
        Zeitdeutung amStandort = BezugsPeriode.zeitpunkt("02.11.2026 07:40", AHRENBERG, null);
        Zeitdeutung nachVorlage = BezugsPeriode.zeitpunkt("02.11.2026 07:40", ZoneId.of("UTC"), null);
        assertThat(amStandort.befund()).isNull();
        assertThat(nachVorlage.befund()).isNull();
        assertThat(BezugsPeriode.iso(amStandort.zeitpunkt(), AHRENBERG)).isEqualTo("2026-11-02T07:40:00+01:00");
        assertThat(amStandort.zeitpunkt())
                .as("dieselbe Wanduhrzeit, eine andere Zone: ein anderer Zeitpunkt")
                .isNotEqualTo(nachVorlage.zeitpunkt());

        // Auch die Periodengrenzen sind Ortszeit: der Oktober beginnt am Standort um 00:00.
        Periodendeutung oktober = BezugsPeriode.periode(
                "31.10.2026",
                null,
                null,
                "periodenende",
                "monat",
                AHRENBERG,
                BezugsPeriode.zeit("2026-11-03T09:12:00+01:00"));
        assertThat(BezugsPeriode.iso(oktober.von(), AHRENBERG)).isEqualTo("2026-10-01T00:00:00+02:00");
        assertThat(BezugsPeriode.iso(oktober.bis(), AHRENBERG)).isEqualTo("2026-11-01T00:00:00+01:00");
    }

    /**
     * IP-3 — es bleibt KEINE zweite Fassung: die Periodendeutung wohnt nur noch hier, und
     * {@link BezugsdatenRegeln} ruft sie an (auch der Stichtag eines Stammdatums).
     */
    @Test
    void diePeriodenlogikStehtNurNochInDiesemModul() throws Exception {
        String regeln = Files.readString(REGELN);
        assertThat(regeln)
                .as("keine zweite Zonen-/Kalenderrechnung in BezugsdatenRegeln")
                .doesNotContain("getValidOffsets")
                .doesNotContain("IsoFields")
                .doesNotContain("MONATSNAMEN");
        assertThat(regeln)
                .as("BezugsdatenRegeln ruft das Modul an")
                .contains("BezugsPeriode.periode(")
                .contains("BezugsPeriode.zeitpunkt(")
                .contains("BezugsPeriode.stundenDesTages(")
                .contains("BezugsPeriode.spanneVon(");

        Instant jetzt = BezugsPeriode.zeit("2026-11-03T09:12:00+01:00");
        assertThat(BezugsdatenRegeln.periode("2026-10", null, null, "periode", "monat", AHRENBERG, jetzt))
                .isEqualTo(BezugsPeriode.periode("2026-10", null, null, "periode", "monat", AHRENBERG, jetzt));
    }
}

File: frontend/portal/src/bezugsPeriode.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it } from 'vitest';
import {
  DATUM_UNLESBAR,
  PERIODE_NICHT_ZU_ENDE,
  PERIODE_PASST_NICHT,
  SAETZE,
  ZEIT_MEHRDEUTIG,
  ZEIT_NICHT_VORHANDEN,
  iso,
  periode,
  satz,
  stundenDesTages,
  zeitpunkt,
} from './bezugsPeriode';

/**
 * Das PERIODEN-Modul (UEMS AP-09 IP-3) gegen die geteilte Vektor-Datei
 * `docs/contracts/v2/bezugsdaten-vectors.json` — Familien `periode`, `zeit`,
 * `stunden` — und gegen die drei Fallen: der Tag mit 23 oder 25 Stunden, der
 * mehrdeutige bzw. nicht existierende Zeitpunkt und der Unterschied zwischen
 * „passt nicht" und „noch nicht zu Ende". Der Java-Zwilling
 * `BezugsPeriodeTest` fährt dieselbe Datei.
 */

const V2 = resolve(process.cwd(), '../../docs/contracts/v2');
const vectors = JSON.parse(readFileSync(resolve(V2, 'bezugsdaten-vectors.json'), 'utf8')) as Record<string, unknown>;

/** Die Zeitzone des Standorts der Referenzfälle (Ahrenberg). */
const AHRENBERG = 'Europe/Berlin';

const befundSaetze = vectors.befund_saetze as Record<string, string>;
const faelle = vectors.cases as Array<Record<string, unknown>>;

const pruefungen = (regel: string) => {
  const gefunden: Array<{ id: string; name: string; ein: Record<string, unknown>; soll: Record<string, unknown> }> = [];
  for (const fall of faelle) {
    for (const p of (fall.pruefungen ?? []) as Array<Record<string, unknown>>) {
      if (p.regel !== regel) continue;
      gefunden.push({
        id: String(fall.id),
        name: String(p.name),
        ein: p.eingang as Record<string, unknown>,
        soll: p.ergebnis as Record<string, unknown>,
      });
    }
  }
  return gefunden;
};

describe('bezugsPeriode — die Familien `periode`, `zeit` und `stunden` der Vektor-Datei', () => {
  for (const f of pruefungen('periode')) {
    it(`${f.id} · periode :: ${f.name}`, () => {
      const ist = periode(
        {
          text: (f.ein.text as string | null) ?? null,
          vonText: (f.ein.von_text as string | null) ?? null,
          bisText: (f.ein.bis_text as string | null) ?? null,
          deutung: f.ein.deutung as string,
          periodeArt: f.ein.periode_art as string,
          jetzt: f.ein.jetzt ? Date.parse(f.ein.jetzt as string) : null,
        },
        AHRENBERG,
      );
      expect(ist.schluessel).toBe((f.soll.schluessel as string | null) ?? null);
      expect(ist.von === null ? null : iso(ist.von, AHRENBERG)).toBe((f.soll.von as string | null) ?? null);
      expect(ist.bis === null ? null : iso(ist.bis, AHRENBERG)).toBe((f.soll.bis as string | null) ?? null);
      expect(ist.stunden).toBe((f.soll.stunden as number | null) ?? null);
      expect(ist.befund).toBe((f.soll.befund as string | null) ?? null);
    });
  }

  for (const f of pruefungen('zeit')) {
    it(`${f.id} · zeit :: ${f.name}`, () => {
      const ist = zeitpunkt(f.ein.text as string, f.ein.zeitzone as string, (f.ein.offset_in_datei as string) ?? null);
      expect(ist.zeitpunkt === null ? null : iso(ist.zeitpunkt, AHRENBERG)).toBe(
        (f.soll.zeitpunkt as string | null) ?? null,
      );
      expect(ist.befund).toBe((f.soll.befund as string | null) ?? null);
      expect(ist.varianten).toEqual(f.soll.varianten);
    });
  }

  for (const f of pruefungen('stunden')) {
    it(`${f.id} · stunden :: ${f.name}`, () => {
      expect(stundenDesTages(f.ein.tag as string, AHRENBERG)).toBe(f.soll.stunden);
    });
  }
});

describe('bezugsPeriode — die drei Fallen', () => {
  it('die Kundensätze wohnen im Modul und sind die des Vertrags', () => {
    for (const [befund, text] of Object.entries(SAETZE)) {
      expect(text).toBe(befundSaetze[befund]);
    }
    expect(Object.keys(SAETZE).sort()).toEqual(
      [PERIODE_PASST_NICHT, PERIODE_NICHT_ZU_ENDE, ZEIT_MEHRDEUTIG, ZEIT_NICHT_VORHANDEN, DATUM_UNLESBAR].sort(),
    );
  });

  it('FALLE 1 — ein Tag hat 23, 24 oder 25 Stunden, nie immer 24', () => {
    expect(stundenDesTages('2026-10-25', AHRENBERG)).toBe(25);
    expect(stundenDesTages('2027-03-28', AHRENBERG)).toBe(23);
    expect(stundenDesTages('2026-12-02', AHRENBERG)).toBe(24);
    // Und die Periode erbt das: der Oktober 2026 hat 745 Stunden, der März 2027 hat 743.
    const jetzt = Date.parse('2026-11-03T09:12:00+01:00');
    expect(periode({ text: '2026-10', deutung: 'periode', periodeArt: 'monat', jetzt }, AHRENBERG).stunden).toBe(745);
    expect(
      periode(
        { text: '2027-03', deutung: 'periode', periodeArt: 'monat', jetzt: Date.parse('2027-04-02T09:00:00+02:00') },
        AHRENBERG,
      ).stunden,
    ).toBe(743);
  });

  it('FALLE 2 — mehrdeutig und nicht vorhanden sind Befunde mit Kundensatz', () => {
    const doppelt = zeitpunkt('25.10.2026 02:30', AHRENBERG, null);
    expect(doppelt.zeitpunkt).toBeNull();
    expect(doppelt.befund).toBe(ZEIT_MEHRDEUTIG);
    expect(doppelt.varianten).toEqual(['2026-10-25T02:30:00+02:00', '2026-10-25T02:30:00+01:00']);
    expect(satz(ZEIT_MEHRDEUTIG)).toBe(
      'Diesen Zeitpunkt gibt es an diesem Tag zweimal (Zeitumstellung). Geben Sie die Zone an.',
    );

    const fehlt = zeitpunkt('28.03.2027 02:30', AHRENBERG, null);
    expect(fehlt.zeitpunkt).toBeNull();
    expect(fehlt.befund).toBe(ZEIT_NICHT_VORHANDEN);
    expect(fehlt.varianten).toEqual([]);
    expect(satz(ZEIT_NICHT_VORHANDEN)).toBe('Diesen Zeitpunkt gibt es an diesem Tag nicht (Zeitumstellung).');

    const mitOffset = zeitpunkt('25.10.2026 02:30', AHRENBERG, '+02:00');
    expect(mitOffset.befund).toBeNull();
    expect(iso(mitOffset.zeitpunkt as number, AHRENBERG)).toBe('2026-10-25T02:30:00+02:00');
  });

  it('FALLE 3 — „passt nicht" ist nicht „noch nicht zu Ende"', () => {
    const jetzt = Date.parse('2026-11-03T09:12:00+01:00');
    const woche = periode(
      { vonText: '28.09.2026', bisText: '04.10.2026', deutung: 'von_bis', periodeArt: 'monat', jetzt },
      AHRENBERG,
    );
    expect(woche.befund).toBe(PERIODE_PASST_NICHT);
    expect(woche.schluessel).toBeNull();
    expect(woche.von).toBeNull();
    expect(satz(PERIODE_PASST_NICHT)).toBe('Der gelieferte Zeitraum ist keine Periode dieser Bezugsgröße.');

    const inWochenreihe = periode(
      { vonText: '28.09.2026', bisText: '04.10.2026', deutung: 'von_bis', periodeArt: 'woche', jetzt },
      AHRENBERG,
    );
    expect(inWochenreihe.schluessel).toBe('2026-W40');
    expect(inWochenreihe.befund).toBeNull();

    const laufend = periode({ text: '2026-11', deutung: 'periode', periodeArt: 'monat', jetzt }, AHRENBERG);
    expect(laufend.befund).toBe(PERIODE_NICHT_ZU_ENDE);
    expect(satz(PERIODE_NICHT_ZU_ENDE)).toBe('Diese Periode ist noch nicht zu Ende.');
    expect(satz(PERIODE_PASST_NICHT)).not.toBe(satz(PERIODE_NICHT_ZU_ENDE));

    // Derselbe Text nach dem Ende der Periode: kein Befund mehr.
    const spaeter = periode(
      { text: '2026-11', deutung: 'periode', periodeArt: 'monat', jetzt: Date.parse('2026-12-01T00:00:00+01:00') },
      AHRENBERG,
    );
    expect(spaeter.befund).toBeNull();
    expect(spaeter.schluessel).toBe('2026-11');
  });

  it('E7/Z5 — die zonenlose Ortszeit liest die Zone des Standorts', () => {
    const amStandort = zeitpunkt('02.11.2026 07:40', AHRENBERG, null);
    const nachVorlage = zeitpunkt('02.11.2026 07:40', 'UTC', null);
    expect(iso(amStandort.zeitpunkt as number, AHRENBERG)).toBe('2026-11-02T07:40:00+01:00');
    expect(amStandort.zeitpunkt).not.toBe(nachVorlage.zeitpunkt);

    const oktober = periode(
      {
        text: '31.10.2026',
        deutung: 'periodenende',
        periodeArt: 'monat',
        jetzt: Date.parse('2026-11-03T09:12:00+01:00'),
      },
      AHRENBERG,
    );
    expect(iso(oktober.von as number, AHRENBERG)).toBe('2026-10-01T00:00:00+02:00');
    expect(iso(oktober.bis as number, AHRENBERG)).toBe('2026-11-01T00:00:00+01:00');
  });

  it('es bleibt keine zweite Fassung: bezugsdaten.ts reicht das Modul weiter', () => {
    const quelle = readFileSync(resolve(process.cwd(), 'src/bezugsdaten.ts'), 'utf8');
    expect(quelle).not.toContain('getValidOffsets');
    expect(quelle).not.toContain('MONATSNAMEN');
    expect(quelle).not.toContain('longOffset');
    expect(quelle).toContain("} from './bezugsPeriode';");
  });
});
